File: src/main/com/ocaml/ide/files/OCamlFileType.java
// Copyright 2000-2020 JetBrains s.r.o. Use of this source code is governed by the Apache 2.0 license that can be found in the LICENSE file.
// http://www.apache.org/licenses/LICENSE-2.0
package com.ocaml.ide.files;

import com.intellij.openapi.fileTypes.LanguageFileType;
import com.ocaml.OCamlBundle;
import com.ocaml.OCamlLanguage;
import com.ocaml.icons.OCamlIcons;
import org.jetbrains.annotations.Contract;
import org.jetbrains.annotations.Nls;
import org.jetbrains.annotations.NotNull;

import javax.swing.*;

/**
 * OCaml source file
 */
public final class OCamlFileType extends LanguageFileType {
    // extension
    public static final String DEFAULT_EXTENSION = "ml";
    public static final String DOT_DEFAULT_EXTENSION = ".ml";

    // instance
    public static final OCamlFileType INSTANCE = new OCamlFileType();

    private OCamlFileType() {
        super(OCamlLanguage.INSTANCE);
    }

    @Contract(pure = true) public static boolean isFile(@NotNull String name) {
        return name.endsWith(DOT_DEFAULT_EXTENSION);
    }

    // implementation

    @Override @NotNull public String getName() {
        return "OCAML";
    }

    @Override @NotNull public String getDescription() {
        return OCamlBundle.message("filetype.ml.description");
    }

    @Override @NotNull public String getDefaultExtension() {
        return DEFAULT_EXTENSION;
    }

    @Override public Icon getIcon() {
        return OCamlIcons.FileTypes.OCAML_SOURCE;
    }

    @Override public @Nls @NotNull String getDisplayName() {
        return getDescription();
    }
}

File: src/main/kotlin/com/ocaml/ide/files/OCamlFileType.kt
package com.ocaml.ide.files

import com.intellij.openapi.fileTypes.LanguageFileType
import com.ocaml.OCamlBundle
import com.ocaml.icons.OCamlIcons
import org.jetbrains.annotations.Contract
import javax.swing.Icon

object OCamlFileType : LanguageFileType(OCamlLanguage) {
    // Extension-Related Constants
    internal const val DEFAULT_EXTENSION = "ml"
    private const val DOT_DEFAULT_EXTENSION = ".$DEFAULT_EXTENSION"

    // LanguageFileType implementation
    override fun getName(): String  = "OCaml File"
    override fun getDescription(): String = OCamlBundle.message("filetype.ml.description")
    override fun getDefaultExtension(): String = DEFAULT_EXTENSION
    override fun getIcon(): Icon = OCamlIcons.FileTypes.OCAML_SOURCE
    override fun getDisplayName(): String  = description

    /**
     * Given a path to an interface, return the path to the source
     *
     * @param path path to the interface file, must be properly formatted
     * @return the path to the source
     * @throws IllegalArgumentException if the path is valid using [OCamlFileType.isFile]
     */
    fun fromInterface(path: String): String {
        if (OCamlInterfaceFileType.isFile(path)) return path.replace(
            OCamlInterfaceFileType.DEFAULT_EXTENSION,
            DEFAULT_EXTENSION
        )
        throw IllegalArgumentException("Not a valid interface file '$path'")
    }
    @Contract(pure = true)
    fun isFile(name: String): Boolean {
        return name.endsWith(DOT_DEFAULT_EXTENSION)
    }
}